tests/FileSystemTests.cpp

```diff
@@ -773,6 +776,66 @@ TEST_CASE("FileSystemTests::CreateFileW helpers", "[filesystem]")
 }
 
 #endif
+TEST_CASE("FileSystemTest::FolderChangeReader destructor does not hang", "[filesystem]")
+{
+    wil::unique_cotaskmem_string testRootTmp;
+    REQUIRE_SUCCEEDED(wil::ExpandEnvironmentStringsW(L"%TEMP%\\wil_test_filesystem", testRootTmp));
+    std::wstring testRootDir = testRootTmp.get();
+    std::wstring testFile = testRootDir + L"\\test.dat";
+    bool deleteDir = false;
+    wil::unique_event_nothrow opCompletedEv; 
+    wil::unique_event_nothrow wilReaderNotify;
+
+    REQUIRE_SUCCEEDED(opCompletedEv.create(wil::EventOptions::None));
+    REQUIRE_SUCCEEDED(wilReaderNotify.create(wil::EventOptions::ManualReset));
+
+    REQUIRE_FALSE(DirectoryExists(testRootDir.c_str()));
+    REQUIRE_SUCCEEDED(wil::CreateDirectoryDeepNoThrow(testRootDir.c_str()));
+    REQUIRE(DirectoryExists(testRootDir.c_str()));
+
+    /**
+     * Move the operation to a new thread. 
+     * The destructor of unique_folder_change_reader might hang. If this happens, 
+     * we want to report an test error instead of hanging forever. 
+     * Initialize the reader in current thread to make sure there is no race condition with test 
+     * creating files
+     */
+    auto reader = wil::make_folder_change_reader_nothrow(testRootDir.c_str(), false, wil::FolderChangeEvents::All,
+    [&](wil::FolderChangeEvent, PCWSTR)
+    {
+        if (deleteDir)
+            RemoveDirectoryW(testRootDir.c_str());
+
+        SetEvent(opCompletedEv.get());
+    });
+    auto readerThread = std::thread([&wilReaderNotify, r = std::move(reader)]() mutable {

```

Comment:
Since the test allows waiting for the thread to time out, probably don't want to capture `wilReaderNotify` by reference. IMO - keep the `HANDLE` stashed locally and then `std::move` the `unique_event` into the lambda so that the thread can be the one to destroy the event.